array_api_strict/_array_object.py

```diff
@@ -149,29 +148,40 @@ def __repr__(self) -> str:
 
     __str__ = __repr__
 
-    # `__array__` was implemented historically for compatibility, and removing it has
-    # caused issues for some libraries (see
-    # https://github.com/data-apis/array-api-strict/issues/67).
-
-    # Instead of `__array__` we now implement the buffer protocol.
-    # Note that it makes array-apis-strict requiring python>=3.12
     def __buffer__(self, flags):
         if self._device != CPU_DEVICE:
-            raise RuntimeError(f"Can not convert array on the '{self._device}' device to a Numpy array.")
+            raise RuntimeError(
+                # NumPy swallows this exception and falls back to __array__.
+                f"Can't extract host buffer from array on the '{self._device}' device."
+            )
         return self._array.__buffer__(flags)
 
-    # We do not define __release_buffer__, per the discussion at
-    # https://github.com/data-apis/array-api-strict/pull/115#pullrequestreview-2917178729
-
-    def __array__(self, *args, **kwds):
-        # a stub for python < 3.12; otherwise numpy silently produces object arrays
-        import sys
-        minor, major = sys.version_info.minor, sys.version_info.major
-        if major < 3 or minor < 12:
+    # `__array__` is not part of the Array API. Ideally we want to support
+    # `xp.asarray(Array)` exclusively through the __buffer__ protocol; however this is
+    # only possible on Python >=3.12. Additionally, when __buffer__ raises (e.g. because
+    # the array is not on the CPU device, NumPy will try to fall back on __array__ but,
+    # if that doesn't exist, create a scalar numpy array of objects which contains the
+    # array_api_strict.Array. So we can't get rid of __array__ entirely.
+    def __array__(
+        self, dtype: None | np.dtype[Any] = None, copy: None | bool = None
+    ) -> npt.NDArray[Any]:
+        if self._device != CPU_DEVICE:
+            # We arrive here from np.asarray() on Python >=3.12 when __buffer__ raises.
+            raise RuntimeError(
+                f"Can't convert array on the '{self._device}' device to a "
+                "NumPy array."
+            )
```

Comment:
Previously, np.asarray() would raise `ValueError: object __array__ method not producing an array` because `__array__` subtly returned None. Changed to explicit behaviour.